pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+                    )
+
+                obj.do_check_domain = False
+
+                # invert the operation if required
+                obj.inverse = info.get("inverse", False)
+
+                if isinstance(obj, STATE_PREP_OPS):
+                    if self._ops:
+                        raise ValueError(
+                            f"State preperation operation {obj} must occur prior to any quantum operations."
+                        )
+
+                    self._prep.append(obj)
+                else:
+                    self._ops.append(obj)
+
+            elif isinstance(obj, MeasurementProcess):
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
+                    self._output_dim += 2 ** len(obj.wires)
+
+                elif "owns" in info:
+                    # TODO: remove the following line once devices
+                    # have been refactored to no longer use obs.return_type.
+                    # Monkeypatch the observable to have the same return
+                    # type as the measurement process.
+                    info["owns"].return_type = obj.return_type
+                    info["owns"].do_check_domain = False
+
+                    self._obs.append((obj, info["owns"]))
+                    self._output_dim += 1
+
+                    if obj.return_type is qml.operation.Sample:
+                        self.is_sampled = True
+
+            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
+                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
+
+        self._update()
+
+    def _update_circuit_info(self):
+        """Update circuit metadata"""
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
+
+    def _update_par_info(self):
+        """Update the parameter information dictionary"""
+        param_count = 0
+
+        for obj in self.operations + self.observables:
+
+            for p in range(len(obj.data)):
+                info = self._par_info.get(param_count, {})
+                info.update({"op": obj, "p_idx": p})
+
+                self._par_info[param_count] = info
+                param_count += 1
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
+
+    def _update(self):
+        """Update all internal tape metadata regarding processed operations and observables"""
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
+
+    def _expand(self, stop_at=None):
+        """Expand all operations and tapes in the processed queue.
+
+        Args:
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
```

Comment:
Currently, they are names e.g., `"CNOT"`. This is because, mainly:

* It works well with the existing device API; since the device API enforces all devices to return a set of strings corresponding to supported operations, you can simply do `tape.expand(stop_at=dev.operations)` for a given device.

Furthermore:

* It doesn't make sense to pass objects, because this means instantiating objects in cases where you might not already have them available (adding overhead)

* We _could_ pass the classes themselves, but this also feel clunky.

* Strings feel pythonic

Reply:
What if you want to stop at one gate, but there's another gate with the same name somewhere else at another level of nesting that you don't want to stop at? How would you differentiate?

Reply:
if they are differentiated by nesting level, I would say this is best differentiated that way.

If they are at the _same_ nesting level, or you explicitly want one gate expanded but another not (regardless of nesting level)... than this is harder. One solution I can think of is if the `stop_at` list contains string _or_ objects:

```python
tape.expand(stop_at=["PauliX", "PauliY", some_op])
```

Is there a use case for this?

Reply:
I've overhauled this. I realised that it is hard to guess exactly how/when a user/developer might want the stopping criterion to look.

Instead, I've made this now a callable, where you can set your own stopping condition as a function of the queue object.

E.g.:

```pycon
>>> tape.expand(stop_at=lambda obj: obj in ["PauliZ", "PauliX"])
>>> tape.expand(stop_at=lambda obj: isinstance(obj, qml.Hadamard))
>>> tape.expand(stop_at=lambda obj: obj is some_op)
```

This should be super flexible, and also similar to some other standard library functions (e.g., the `key` argument for `sorted`).

Reply:
Nice!